File: apps/backend/src/vivecaribe/api/routers/reports.py
"""Dashboard report HTTP routes."""

from __future__ import annotations

from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query

from vivecaribe.api.deps import CurrentUser, ReportsRepo
from vivecaribe.domain.enums import BookingProvider
from vivecaribe.domain.reports import (
    DemographicItem,
    MonthlySalesPoint,
    MonthlyStatisticsPoint,
    ProviderCardItem,
    ProximoPartidoItem,
    ReportSummary,
    TopCityItem,
    TopProviderItem,
)

router = APIRouter(prefix="/reports", tags=["reports"])


@router.get("/summary", response_model=ReportSummary)
async def report_summary(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[
        date | None,
        Query(description="Inclusive lower bound (America/Bogota calendar day)"),
    ] = None,
    fecha_to: Annotated[
        date | None,
        Query(description="Inclusive upper bound (America/Bogota calendar day)"),
    ] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
) -> ReportSummary:
    """Return dashboard KPI cards."""
    return await reports.get_summary(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
    )


@router.get("/statistics", response_model=list[MonthlyStatisticsPoint])
async def report_statistics(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
) -> list[MonthlyStatisticsPoint]:
    """Return monthly income_final vs costos series."""
    return await reports.get_statistics(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
    )


@router.get("/monthly-sales", response_model=list[MonthlySalesPoint])
async def report_monthly_sales(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
) -> list[MonthlySalesPoint]:
    """Return monthly profit series."""
    return await reports.get_monthly_sales(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
    )


@router.get("/top-providers", response_model=list[TopProviderItem])
async def report_top_providers(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 10,
) -> list[TopProviderItem]:
    """Return booking providers ranked by profit."""
    return await reports.get_top_providers(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
        limit=limit,
    )


@router.get("/top-cities", response_model=list[TopCityItem])
async def report_top_cities(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 10,
) -> list[TopCityItem]:
    """Return experience cities ranked by profit."""
    return await reports.get_top_cities(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
        limit=limit,
    )


@router.get("/demographics", response_model=list[DemographicItem])
async def report_demographics(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=50)] = 10,
) -> list[DemographicItem]:
    """Return visitor countries ranked by participants."""
    return await reports.get_demographics(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
        limit=limit,
    )


@router.get("/providers", response_model=list[ProviderCardItem])
async def report_providers(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    fecha_from: Annotated[date | None, Query()] = None,
    fecha_to: Annotated[date | None, Query()] = None,
    booking_provider: Annotated[BookingProvider | None, Query()] = None,
) -> list[ProviderCardItem]:
    """Return per-provider card summaries."""
    return await reports.get_providers(
        fecha_from=fecha_from,
        fecha_to=fecha_to,
        booking_provider=booking_provider,
    )


@router.get("/proximos-partidos", response_model=list[ProximoPartidoItem])
async def report_proximos_partidos(
    reports: ReportsRepo,
    _current_user: CurrentUser,
    limit: Annotated[int, Query(ge=1, le=20)] = 5,
) -> list[ProximoPartidoItem]:
    """Return the next upcoming partidos (ignores dashboard filters)."""
    return await reports.get_proximos_partidos(limit=limit)

File: apps/frontend/src/lib/reports.ts
import { apiJson } from "@/lib/api";
import type { BookingProvider } from "@/types/reservation";
import type { PartidoListItem } from "@/types/partido";

export type ReportFilters = {
  fecha_from?: string;
  fecha_to?: string;
  booking_provider?: string;
};

export type ReportSummary = {
  participants: number;
  reservas: number;
  profit: string;
  partidos: number;
};

export type MonthlyStatisticsPoint = {
  month: string;
  income_final: string;
  costos: string;
};

export type MonthlySalesPoint = {
  month: string;
  profit: string;
};

export type TopProviderItem = {
  provider: BookingProvider | string;
  profit: string;
  reservas: number;
  participants: number;
};

export type TopCityItem = {
  city: string;
  profit: string;
  reservas: number;
};

export type TemporadaPoint = {
  month: string;
  city: string;
  participants: number;
};

export type DemographicItem = {
  country: string;
  reservas: number;
  participants: number;
};

export type ProviderCardItem = {
  provider: BookingProvider | string;
  profit: string;
  reservas: number;
  participants: number;
};

function buildReportQuery(filters: ReportFilters = {}): string {
  const query = new URLSearchParams();
  if (filters.fecha_from) query.set("fecha_from", filters.fecha_from);
  if (filters.fecha_to) query.set("fecha_to", filters.fecha_to);
  if (filters.booking_provider) {
    query.set("booking_provider", filters.booking_provider);
  }
  const serialized = query.toString();
  return serialized ? `?${serialized}` : "";
}

export async function fetchReportSummary(
  filters: ReportFilters = {},
): Promise<ReportSummary> {
  return apiJson<ReportSummary>(`/reports/summary${buildReportQuery(filters)}`);
}

export async function fetchReportStatistics(
  filters: ReportFilters = {},
): Promise<MonthlyStatisticsPoint[]> {
  return apiJson<MonthlyStatisticsPoint[]>(
    `/reports/statistics${buildReportQuery(filters)}`,
  );
}

export async function fetchReportMonthlySales(
  filters: ReportFilters = {},
): Promise<MonthlySalesPoint[]> {
  return apiJson<MonthlySalesPoint[]>(
    `/reports/monthly-sales${buildReportQuery(filters)}`,
  );
}

export async function fetchReportTopProviders(
  filters: ReportFilters = {},
): Promise<TopProviderItem[]> {
  return apiJson<TopProviderItem[]>(
    `/reports/top-providers${buildReportQuery(filters)}`,
  );
}

export async function fetchReportTopCities(
  filters: ReportFilters = {},
): Promise<TopCityItem[]> {
  return apiJson<TopCityItem[]>(
    `/reports/top-cities${buildReportQuery(filters)}`,
  );
}

export async function fetchReportTemporada(
  filters: ReportFilters = {},
): Promise<TemporadaPoint[]> {
  return apiJson<TemporadaPoint[]>(
    `/reports/temporada${buildReportQuery(filters)}`,
  );
}

export async function fetchReportDemographics(
  filters: ReportFilters = {},
): Promise<DemographicItem[]> {
  return apiJson<DemographicItem[]>(
    `/reports/demographics${buildReportQuery(filters)}`,
  );
}

export async function fetchReportProviders(
  filters: ReportFilters = {},
): Promise<ProviderCardItem[]> {
  return apiJson<ProviderCardItem[]>(
    `/reports/providers${buildReportQuery(filters)}`,
  );
}

export async function fetchReportProximosPartidos(): Promise<PartidoListItem[]> {
  return apiJson<PartidoListItem[]>("/reports/proximos-partidos");
}
